src/cli/setup_mode.rs

```diff
@@ -116,39 +96,28 @@ pub fn main() -> Result<utils::ExitCode> {
         }
         Err(e) => return Err(e.into()),
     };
-    let no_prompt = matches.get_flag("no-prompt");
-    let verbose = matches.get_flag("verbose");
-    let quiet = matches.get_flag("quiet");
-    let default_host = matches
-        .get_one::<String>("default-host")
-        .map(ToOwned::to_owned);
-    let default_toolchain = matches
-        .get_one::<MaybeOfficialToolchainName>("default-toolchain")
-        .map(ToOwned::to_owned);
-    let profile = matches
-        .get_one::<String>("profile")
-        .expect("Unreachable: Clap should supply a default");
-    let no_modify_path = matches.get_flag("no-modify-path");
-    let no_update_toolchain = matches.get_flag("no-update-default-toolchain");
-
-    let components: Vec<_> = matches
-        .get_many::<String>("components")
-        .map(|v| v.map(|s| &**s).collect())
-        .unwrap_or_else(Vec::new);
-
-    let targets: Vec<_> = matches
-        .get_many::<String>("targets")
-        .map(|v| v.map(|s| &**s).collect())
-        .unwrap_or_else(Vec::new);
+
+    if self_replace {
+        return self_update::self_replace();
+    }
+
+    if dump_testament {
+        common::dump_testament()?;
+        return Ok(utils::ExitCode(0));
+    }
 
     let opts = InstallOpts {
         default_host_triple: default_host,
-        default_toolchain,
+        default_toolchain: default_toolchain.map(|it| it.to_owned()),
         profile: profile.to_owned(),
         no_modify_path,
-        no_update_toolchain,
-        components: &components,
-        targets: &targets,
+        no_update_toolchain: no_update_default_toolchain,
+        components: &components
+            .iter()
+            .flatten()
+            .map(|s| &**s)
+            .collect::<Vec<_>>(),
+        targets: &targets.iter().flatten().map(|s| &**s).collect::<Vec<_>>(),
```

Comment:
This is indeed a bit awkward.
I think this change it's a bit OT but I guess it's okay to add a tiny refactoring to this PR though, if the size of this change remains relatively small.
Let me have a look...

Reply:
@djc I've changed my mind... It seems that this `InstallOpts` is just passing this `&[&str]` down to other functions. I think it's quite standard to have `&[&str]` as function arguments, although this would mean that the `|s| &**s` dance will certainly happen in one way or another.. (I do this in my project as well, although I prefer `AsRef::as_ref`.)